go-controller/pkg/ovn/gateway_cleanup.go

```diff
@@ -159,3 +159,96 @@ func staticRouteCleanup(nextHops []net.IP) {
 		}
 	}
 }
+
+// multiJoinSwitchGatewayCleanup removes all the OVN NB gateway logical entities created based
+// on the obsoleted multiple join switch OVN topology. This function is used for upgrading OVN
+// logical topology to single join switch version.
+func multiJoinSwitchGatewayCleanup(nodeName string) error {
+	gatewayRouter := gwRouterPrefix + nodeName
+
+	// Get the gateway router port's IP address (connected to join switch)
+	var nextHops []net.IP
+
+	gwIPAddrs, err := util.GetLRPAddrs(gwRouterToJoinSwitchPrefix + gatewayRouter)
+	if err != nil {
+		return err
+	}
+
+	for _, gwIPAddr := range gwIPAddrs {
+		// Delete logical router policy whose nexthop is the old rtoj- gateway port address
+		stdout, stderr, err := util.RunOVNNbctl("--data=bare", "--no-heading", "--columns=_uuid",
+			"find", "logical_router_policy", fmt.Sprintf("nexthop=%s", gwIPAddr.IP))
+		if err != nil {
+			klog.Errorf("Unable to find LR policy of nexthop: %s for node %s, stderr: %s, err: %v",
+				gwIPAddr.IP, nodeName, stderr, err)
+			continue
+		}
+		if stdout != "" {
+			policyIDs := strings.Fields(stdout)
+			for _, policyID := range policyIDs {
+				_, stderr, err = util.RunOVNNbctl("remove", "logical_router", ovnClusterRouter, "policies", policyID)
+				if err != nil {
+					klog.Errorf("Unable to remove LR policy: %s, stderr: %s, err: %v", policyID, stderr, err)
+				}
+			}
+		}
+		nextHops = append(nextHops, gwIPAddr.IP)
+	}
+	staticRouteCleanup(nextHops)
+
+	// Remove the join switch that connects ovn_cluster_router to gateway router
+	_, stderr, err := util.RunOVNNbctl("--if-exist", "ls-del", "join_"+nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to delete the join logical switch %s, "+
+			"stderr: %q, error: %v", "join_"+nodeName, stderr, err)
+	}
+
+	// Remove the gateway router associated with nodeName
```

Comment:
why do we need to delete the logical router? or the external switch? Only thing that needs to be disconnected is the join switch, then the link between join switch and the GR reconfigured.

Reply:
same as above. Thanks